### PR TITLE
669 Add abort reason to runWithRetry

### PR DESCRIPTION
**Pull request by kian-thompson**
# [AB#669](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/669)

[How contribute to this repo](https://github.com/microsoft/FluidFramework/blob/main/CONTRIBUTING.md).

[Guidelines for Pull Requests](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

## Description

We've seen a few cases where a host bug manifests as triggering an error with message "runWithRetry was Aborted". This message doesn't give any hint as to which API was being called when it was aborted, or the reason for the cancellation.

This PR provides a new-and-improved version of runWithRetry that returns a result (failed, or aborted) instead of throwing when something goes wrong, allowing us to move the error instrumentation to the caller (Container.attach() in this case), where we can put the appropriate level of detail in the error message.

See https://github.com/microsoft/FluidFramework/pull/11000

## Upcoming breaking change

This PR will have an upcoming breaking change which replaces the implementation of `runWithRetry` with the new `runWithRetry2`. The new method will be removed at the same time this change is made.
See https://dev.azure.com/fluidframework/internal/_workitems/edit/3042

### CODE REVIEW

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts [hunk 1]
```diff
@@ -104,26 +106,31 @@ export class RetriableDocumentStorageService implements IDocumentStorageService,
...
             },
         );
+
+        switch (runResult.status) {
+            case "succeeded":
+                return runResult.result;
+            case "failed":
+                throw runResult.error;
+            case "aborted": {
+                this.logger.sendTelemetryEvent({
```
Comment (kian-thompson) [inline packages/loader/container-loader/src/retriableDocumentStorageService.ts:125]: Possible TODO: https://github.com/microsoft/FluidFramework/pull/9761/files

File: packages/loader/driver-utils/src/runWithRetry.ts [hunk 2]
```diff
@@ -107,3 +109,88 @@ export async function runWithRetry<T>(
...
 }
+
+export type RunResult<T> = {
+    status: "succeeded";
+    result: T;
+} | {
+    status: "failed";
+    error: unknown;
+} | {
+    status: "aborted";
```
Comment (anthony-murphy) [inline packages/loader/driver-utils/src/runWithRetry.ts:120]: can or should aborted have a reason? maybe optional?

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts [hunk 3]
```diff
@@ -104,26 +106,31 @@ export class RetriableDocumentStorageService implements IDocumentStorageService,
...
+                    eventName: `${callName}_abortedStorageDisposed`,
+                    fetchCallName: callName, // fetchCallName matches logs in runWithRetry.ts
+                });
+                // pre-0.58 error message: storageServiceDisposedCannotRetry
+                throw new GenericError("Storage Service is disposed. Cannot retry", { canRetry: false });
```
Comment (vladsud) [inline packages/loader/container-loader/src/retriableDocumentStorageService.ts:130]: I think you do not need { canRetry: false} here as it's implied. That said, the other place (attach() call) does not have this property, and differences like that RE handling abort signals worry me - I'd rather see it being done uniformly by runWithRetry() code itself

**Comment by vladsud** [comment]
@kian-thompson, the approach you took is more explicit, but also results in more code written.
I wonder if you considered passing the reason to AbortController.abort (see https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort & https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/reason), and having runWithRetry() through properly formatted error object where callers do not need to do anything.

I believe I've heard from someone that maybe this extra arg is not supported in Node or something like that, but even if that's the case, it feels like it's easy to remedy.

**Comment by vladsud** [comment]
This is likely bonus point, and possibly needs to be handled in separate task / PR: It's worth taking a look at ODSP implementation of runWithRetry. While its shape and implementation differ a lot, looking at call sites I see code that will eventually need to pass AbortySignal through the stack too (eventually), and thus it will need similar handling for abort flow. Is there a way to reuse at least some of the logic here?

**Comment by kian-thompson** [comment]
> This is likely bonus point, and possibly needs to be handled in separate task / PR: It's worth taking a look at ODSP implementation of runWithRetry. While its shape and implementation differ a lot, looking at call sites I see code that will eventually need to pass AbortySignal through the stack too (eventually), and thus it will need similar handling for abort flow. Is there a way to reuse at least some of the logic here?

[AB#2994](https://dev.azure.com/fluidframework/internal/_workitems/edit/2994)

File: BREAKING.md [hunk 4]
```diff
@@ -48,6 +50,9 @@ The following legacy events and event parameters have been marked as deprecated
         - This event has been unused and unsupported for some time.
           No replacement API recommended.
 
+### Upcoming change to `runWithRetry`
```
Comment (vladsud) [inline BREAKING.md:53]: Do we have a follow up ticket to track this work?

Comment (kian-thompson) [reply BREAKING.md:53]: [AB#3042](https://dev.azure.com/fluidframework/internal/_workitems/edit/3042)

**Pull request closed without merging**